### PR TITLE
ADR Proposal: Add React Framework

### PR DESCRIPTION
**Pull request by maxxcrawford**
# References: 
Jira: [MNTOR-1588](https://mozilla-hub.atlassian.net/browse/MNTOR-1588)

# Description
Add ADR for proposed migration to React front-end framework 

# TODO:
- [ ] Add sample code (?) 
- [ ] Add Migration Plan Proposal
- [ ] Add secondary decisions that would need to be considered/answered from this decision

# Checklist (Definition of Done)
- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [ ] All acceptance criteria are met.
- [ ] ~Localization strings (if needed) have been added.~
- [ ] ~I've added or updated the relevant sections in readme and/or code comments~
- [ ] ~I've added a unit test to test for potential regressions of this bug.~
- [ ] ~Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.~
- [ ] ~Jira ticket has been updated (if needed) to match changes made during the development process.~
- [ ] ~Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.~

### CODE REVIEW

File: docs/adr/0004-framework.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+The Monitor team is tasked with launching a premium product offering as soon as possible. The current architecture of Monitor is not suited for synchronous feature development and lacks the complex client-side state management premium will require. 
+
+
+## Decision Drivers <!-- optional -->
+
+* The expansion in Monitor premium requires a framework to manage increased complexity and client-side state management
+* Promote component sharing between S&P web-based products, Relay & Monitor
+* Allow developers to work on specific components and features without affecting the rest of the codebase (fewer merge conflicts, makes it easier break down tickets)
+* Unidirectional (one-way) data flow makes it easier to understand the flow of data in the application, which may be necessary when handling multiple states (especially for the data removal sequence)
+* Monitor Front-end engineers have extensive React experience
```
Comment (maxxcrawford) [inline docs/adr/0004-framework.md:19]: Extra context: @codemist @Vinnl and myself have valuable experience from when we decoupled the frontend on Relay from Django to React.

File: docs/adr/0004-framework.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	* Note that FxA came to a similiar decision: [FxA ADR: Refactor Subscription Platform frontend - with Next.js](https://github.com/mozilla/fxa/blob/main/docs/adr/0035-refactor-payments-frontend-with-nextjs.md) 
+* Good, Typescript: TypeScript's static typing enhances code robustness and readability by catching errors at compile-time and making code more self-explanatory.
+* Good, Industry Usage/Standard: React has a large and active community of developers, with many resources available, including libraries, tutorials, and support.
+* Good, Simpler Code: Hooks in React utilizes simpler code that implements similar functionalities faster and more effectively
+* Good, [fluent-react](https://github.com/projectfluent/fluent.js/tree/main/fluent-react) enables more dynamic localized texts and messages
+* Bad, Tooling Overhead: React's tooling and ecosystem can be complex, which can be a disadvantage for developers who prefer a more minimalistic approach.
+* Bad, Performance cost when rendering a large number of components, may require additional optimization work
```
Comment (mansaj) [inline docs/adr/0004-framework.md:46]: I would also say that in terms of overall site performance, it's a net positive when server side calls are returning json objects instead of html, doing react well can reduce roundtrip time and frequency

File: docs/adr/0004-framework.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# Use React Framework
```
Comment (mansaj) [inline docs/adr/0004-framework.md:1]: I would change the file name from "0004-framework" to something like "0004-frontend-framework" in order to diff with potential future ADRs about other types of framework decisions.

File: docs/adr/0004-framework.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+* Bad, No Virtual DOM: Current architecture may be slower than React, as it lacks the optimizations provided by a virtual DOM.
+
+### Use a different framework
+
+* Bad, Additional tool fracturing across the Mozilla ecosystem
+
+## Links <!-- optional -->
+
+* [React usage, State of JavaScript](https://2022.stateofjs.com/en-US/libraries/front-end-frameworks/)
+* [Previous ADR for Native Templating](https://github.com/mozilla/blurts-server/blob/main/docs/adr/0001-native-templating.md)
```
Comment (toufali) [inline docs/adr/0004-framework.md:68]: Additional link for extensive context in deciding framework or no:
https://javarome.medium.com/design-noframework-bbc00a02d9b3

Spoiler alert - It skews no, but includes points in favor of frameworks when appropriate.

File: docs/adr/0004-framework.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+* Good, Shared knowledge: This approach shares the same structure as Firefox Relay. This would immediately allow the team to reuse and reference.
+	* Note that FxA came to a similiar decision: [FxA ADR: Refactor Subscription Platform frontend - with Next.js](https://github.com/mozilla/fxa/blob/main/docs/adr/0035-refactor-payments-frontend-with-nextjs.md) 
+* Good, Typescript: TypeScript's static typing enhances code robustness and readability by catching errors at compile-time and making code more self-explanatory.
+* Good, Industry Usage/Standard: React has a large and active community of developers, with many resources available, including libraries, tutorials, and support.
+* Good, Simpler Code: Hooks in React utilizes simpler code that implements similar functionalities faster and more effectively
+* Good, [fluent-react](https://github.com/projectfluent/fluent.js/tree/main/fluent-react) enables more dynamic localized texts and messages
+* Bad, Tooling Overhead: React's tooling and ecosystem can be complex, which can be a disadvantage for developers who prefer a more minimalistic approach.
```
Comment (Vinnl) [inline docs/adr/0004-framework.md:45]: More pros:

- Avoid snowflake configs: we won't have to manually set up tooling and make sure they work well together, like https://github.com/mozilla/blurts-server/pull/3003 and https://github.com/mozilla/blurts-server/pull/2987.
- Automatic code splitting, avoiding e.g. cumulative layout shift, and decreating bundle sizes.

More cons:
- We give away some autonomy w.r.t. what tooling we can use and when we can use the latest versions.
- The client-side execution time of React is additional overhead.

File: docs/adr/0004-framework.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+*Note: This recommendation is to use Next.js for SSR page generation. We would continue to use Express to handle backend logic, API, etc.*
+
+
+### Leave as-is
+
+* Good, the current website has gone through extensive testing, fox fooding and is working
+* Good, Little Third-Party dependencies
+* Good, No throw away code
+* Bad, Testing: There are currently no front-end specific tests
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply docs/adr/0004-framework.md:58]: Theoretically, yes, but we'd need to both run the entire website, and fire up a full browser (which is [a major source of flakiness](https://testing.googleblog.com/2017/04/where-do-our-flaky-tests-come-from.html#gdcalert5), in addition to being slow to run). It's also a hassle to setup, as evidenced by the fact that we haven't done that :P

File: docs/adr/0004-framework.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+* The expansion in Monitor premium requires a framework to manage increased complexity and client-side state management
+* Promote component sharing between S&P web-based products, Relay & Monitor
+* Allow developers to work on specific components and features without affecting the rest of the codebase (fewer merge conflicts, makes it easier break down tickets)
+* Unidirectional (one-way) data flow makes it easier to understand the flow of data in the application, which may be necessary when handling multiple states (especially for the data removal sequence)
+* Monitor Front-end engineers have extensive React experience
+
+## Considered Options
+
+1. Migrate framework to React and Next.js
+1. Leave as-is
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply docs/adr/0004-framework.md:24]: Great call-out. I linked to the bottom of the doc about the previous decision that encompasses this: https://github.com/mozilla/blurts-server/blob/main/docs/adr/0001-native-templating.md

File: docs/adr/0004-framework.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+* Bad, Performance cost when rendering a large number of components, may require additional optimization work
+* Bad, Learning curve: Challenging for new developers jumping into the project to understand how to debug issues or how components are interacting with each other
+* Bad, Time Cost: Time spent porting existing site over to React
+
+*Note: This recommendation is to use Next.js for SSR page generation. We would continue to use Express to handle backend logic, API, etc.*
+
+
+### Leave as-is
+
+* Good, the current website has gone through extensive testing, fox fooding and is working
```
Comment (Vinnl) [inline docs/adr/0004-framework.md:55]: Additional pros:

- Conveys our support for standards and the web standards process

Additional cons:
- Harder to learn our snowflake framework (e.g. our use of a proxy on one page to update the DOM)
- Easy for server and client-side state to get out of sync
- Hard to re-use code on the client-side (see: dynamic Fluent strings)
- Hard to statically analyse, and in general doesn't make use of a lot of tooling that can help us deliver better quality code without us manually setting them all up (correctly).

File: docs/adr/0004-framework.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+* Good, Little Third-Party dependencies
+* Good, No throw away code
+* Bad, Testing: There are currently no front-end specific tests
+* Bad, No Virtual DOM: Current architecture may be slower than React, as it lacks the optimizations provided by a virtual DOM.
+
+### Use a different framework
+
+* Bad, Additional tool fracturing across the Mozilla ecosystem
```
Comment (Vinnl) [inline docs/adr/0004-framework.md:63]: Additional cons:

- Both Next.js and React have lots of momentum and unlikely to go away anytime soon, and have a history of maintaining backwards compatibility.
- Time spent analysing all potential options is time not spent creating value :)

**Review by Vinnl — APPROVED** [review]
Added some suggestions and pros and cons off the top of my head, to stimulate more discussion (in particular about what to do for the back-end).

File: docs/adr/0004-framework.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
+# Use React Framework
+
+* Status: proposed
+* Deciders: Monitor team
+* Date: 2022-05-03
+* Technical Story: [MNTOR-1588](https://mozilla-hub.atlassian.net/browse/MNTOR-1588)
+
+## Context and Problem Statement
+
+The Monitor team is tasked with launching a premium product offering as soon as possible. The current architecture of Monitor is not suited for synchronous feature development and lacks the complex client-side state management premium will require.
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply docs/adr/0004-framework.md:10]: You got it. 

The quick context is we are going to have six engineers collaborate on the same view/files concurrently (in the current format). There will be a high risk of merge collisions and increased overhead due to conflicting changes, and something the other engineers highlighted as an issue in the current architecture. 

This issue is compounded further with the design/technical requirements that changed to merge the two dashboard views (breach resolution and data removal) into a singular view.

File: docs/adr/0004-framework.md (hunk 4 above)
Comment (maxxcrawford) [reply docs/adr/0004-framework.md]: Added link in 1714708a46edf08c1eea91fa751ddd86dbbe5cac

File: docs/adr/0004-framework.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply docs/adr/0004-framework.md]: @Vinnl @toufali I revised the comment on testing here to reflect on the setup challenge, rather than the lack of it. 

Revised in 1714708a46edf08c1eea91fa751ddd86dbbe5cac

File: docs/adr/0004-framework.md (hunk 8 above)
Comment (maxxcrawford) [reply docs/adr/0004-framework.md:106]: Added in 1714708a46edf08c1eea91fa751ddd86dbbe5cac

File: docs/adr/0004-framework.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+* Good, Industry Usage/Standard: React has a large and active community of developers, with many resources available, including libraries, tutorials, and support.
+* Good, Simpler Code: Hooks in React utilizes simpler code that implements similar functionalities faster and more effectively
+* Good, [fluent-react](https://github.com/projectfluent/fluent.js/tree/main/fluent-react) enables more dynamic localized texts and messages
+* Bad, Tooling Overhead: React's tooling and ecosystem can be complex, which can be a disadvantage for developers who prefer a more minimalistic approach.
+* Bad, Performance cost when rendering a large number of components, may require additional optimization work
+* Bad, Learning curve: Challenging for new developers jumping into the project to understand how to debug issues or how components are interacting with each other
+* Bad, Time Cost: Time spent porting existing site over to React
+
+*Note: This recommendation is to use Next.js for SSR page generation. We would continue to use Express to handle backend logic, API, etc.*
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply docs/adr/0004-framework.md]: Added in 1714708a46edf08c1eea91fa751ddd86dbbe5cac. Note that I left out the "What to do" here, but added a line in the _Decision Outcome_ section to also decide on migration if React is selected.

File: docs/adr/0004-framework.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+* Good, Typescript: TypeScript's static typing enhances code robustness and readability by catching errors at compile-time and making code more self-explanatory.
+* Good, Industry Usage/Standard: React has a large and active community of developers, with many resources available, including libraries, tutorials, and support.
+* Good, Simpler Code: Hooks in React utilizes simpler code that implements similar functionalities faster and more effectively
+* Good, [fluent-react](https://github.com/projectfluent/fluent.js/tree/main/fluent-react) enables more dynamic localized texts and messages
+* Bad, Tooling Overhead: React's tooling and ecosystem can be complex, which can be a disadvantage for developers who prefer a more minimalistic approach.
+* Bad, Performance cost when rendering a large number of components, may require additional optimization work
+* Bad, Learning curve: Challenging for new developers jumping into the project to understand how to debug issues or how components are interacting with each other
+* Bad, Time Cost: Time spent porting existing site over to React
+
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply docs/adr/0004-framework.md:101]: Added items in 1714708a46edf08c1eea91fa751ddd86dbbe5cac

File: docs/adr/0004-framework.md (hunk 5 above)
Comment (maxxcrawford) [reply docs/adr/0004-framework.md:48]: Added in 1714708a46edf08c1eea91fa751ddd86dbbe5cac

File: docs/adr/0004-framework.md (hunk 3 above)
Comment (maxxcrawford) [reply docs/adr/0004-framework.md:1]: @mansaj Great call-out. Adding this as a last step TODO as I don't wanna lose any of the current comments' context.

File: docs/adr/0004-framework.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+* Good, Little Third-Party dependencies
+* Good, No throw away code
+* Good, Conveys our support for standards and the web standards process
+* Bad, Front-end testing requires additional set-up (using a full browser) which would impact test speeds and contribute to flakiness.
+* Bad, No Virtual DOM: Current architecture may be slower than React, as it lacks the optimizations provided by a virtual DOM.
+* Bad, Harder to learn our snowflake framework (e.g. our use of a proxy on one page to update the DOM)
+* Bad, Easy for server and client-side state to get out of sync
+* Bad, Hard to re-use code on the client-side (see: dynamic Fluent strings)
+* Bad, Hard to statically analyse, and in general doesn't make use of a lot of tooling that can help us deliver better quality code without us manually set up
+* Bad, Poor front-end developer experience: Writing in template literals removes tag coloring, auto-complete functions, etc.
```
Comment (maxxcrawford) [inline docs/adr/0004-framework.md:117]: This is a real, daily paper cut. 

_If there's a solve now anyone has, please share it. I've heard from other devs about this and experienced it first hand._ 

HTML inside JavaScript as a string: 👎 
image

Comment (Vinnl) [reply docs/adr/0004-framework.md:117]: Hmm, there's this tool called [Handlebars](https://handlebarsjs.com/) that has widespread editor support? :P

Comment (maxxcrawford) [reply docs/adr/0004-framework.md:117]: 😭 😭 😭

File: docs/adr/0004-framework.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+*Note: This recommendation is to use Next.js for SSR page generation. We would continue to use Express to handle backend logic, API, etc.*
+
+
+### Leave as-is
+
+* Good, the current website has gone through extensive testing, fox fooding and is working
+* Good, Little Third-Party dependencies
+* Good, No throw away code
+* Bad, Testing: There are currently no front-end specific tests
+* Bad, No Virtual DOM: Current architecture may be slower than React, as it lacks the optimizations provided by a virtual DOM.
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply docs/adr/0004-framework.md]: Whoops! Missed this. Removed in  5b1a412a4c45e948119fc3821c16952bb9d68491

File: docs/adr/0004-framework.md (hunk 2 above)
Comment (maxxcrawford) [reply docs/adr/0004-framework.md:50]: @mansaj I want to capture your comment in the doc. How would you propose we re-word this?

File: docs/adr/0004-framework.md (hunk 13 above)
Comment (toufali) [reply docs/adr/0004-framework.md:116]: @maxxcrawford I use the Template Literals VSCode extension: https://marketplace.visualstudio.com/items?itemName=julienetie.vscode-template-literals

Works very well for me for template literal code, but I sometimes need to disable it when working in non-template JS. I had been meaning to add this extension to the recommended tools section of ReadMe, but never got around to it.

Comment (toufali) [reply docs/adr/0004-framework.md:116]: @maxxcrawford there might be other syntax highlight extensions as well, please share if you find a better one!  Regardless, I hope the one above alleviates the added "con" of no syntax highlight.

File: docs/adr/0004-framework.md (hunk 2 above)
Comment (mansaj) [reply docs/adr/0004-framework.md:50]: @maxxcrawford No worries, maybe just something like:
* Good, Reduce network roundtrip time and frequency, improves app performance

**Comment by pdehaan** [comment]
Not sure I have a horse in this race, but&hellip; personally a huge fan of the Next.js + React + TypeScript trident. 🔱 

**PROS:** Lots of other people internally (and on S&P) have experience w/ at least 1-3 of those. Monitor v2 already has pretty solid JSDoc and type definitions, so hopefully migration to React/TSX isn't too bad.

**CONS:** A bit of a learning curve if you aren't familiar w/ Next and/or TypeScript types, but thankfully TypeScript typing is easy to opt in/out of and slowly migrate over to.

File: docs/adr/0004-framework.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lesleyjanenorton) [reply docs/adr/0004-framework.md]: +1 to webdriver tests being flaky/just generally painful. We experienced this in previous incarnations of VPN auth tests and currently struggling with webdriver related woes on the secure-proxy extension.

[the remark this answers is not in the document]
Comment (pdehaan) [reply docs/adr/0004-framework.md]: I defer to @mozrokafor and @jrbenny35, but I think we've been pretty happy w/ [Playwright](https://playwright.dev/) lately as an e2e framework. Although if you need to test _add-ons_, I believe your options are much more limited.
And I'm not sure it makes sense to use Playwright for the front end tests and switch to Selenium for add-on testing (if that is a concern for Monitor).

/selenium enters the chat

[the remark this answers is not in the document]
Comment (Vinnl) [reply docs/adr/0004-framework.md]: We already have Playwright for e2e tests on Monitor. It's indeed more stable than plain Webdriver/Selenium, but it's still firing up a full browser. IMHO this is fine for high-level smoke tests that catch e.g. entire pages disappearing or the database connection being dropped, but for unit tests (where you'd probably also want to use something like [Web Test Runner](https://modern-web.dev/docs/test-runner/overview/) or [Karma](https://karma-runner.github.io) to run isolated units of code), it's still a lot flakier and slower than running in pure Node.

[the remark this answers is not in the document]
Comment (pdehaan) [reply docs/adr/0004-framework.md]: Absolutely. Playwright excels at e2e, but I wouldn't recommend it for unit tests. On FxA we currently use Jest and Storybook for front end testing. https://github.com/search?q=repo%3Amozilla%2Ffxa%20jest&type=code
Also https://kentcdodds.com/blog/write-tests

And somewhere in the bowels of the CircleCI or GitHub Actions configs we have some script that deploys the Storybook builds to S3/GCP and gets linked from each PR.
https://github.com/mozilla/fxa/blob/11d541f26313bdef8050ff59bd14910c070c6e70/.circleci/config.yml#L686-L694

File: docs/adr/0004-framework.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+### Migrate framework to React and Next.js
+
+* Good, Component-based Architecture: This simplifies code maintenance by breaking complex UIs into reusable components.
+* Good, Testing: the virtual DOM makes it feasible to write non-brittle (i.e. without needing to fire up a full browser) unit tests of the front-end
+* Good, Fewer bugs: having React take care of removing, adding or changing DOM nodes removes a large surface area for bugs.
+* Good, Shared knowledge: This approach shares the same structure as Firefox Relay. This would immediately allow the team to reuse and reference.
+	* Note that FxA came to a similiar decision: [FxA ADR: Refactor Subscription Platform frontend - with Next.js](https://github.com/mozilla/fxa/blob/main/docs/adr/0035-refactor-payments-frontend-with-nextjs.md) 
+* Good, Enables static analysis: programmatically-generated DOM and explicit module imports enables optimal use of TypeScript to catch errors at compile-time and make code more self-explanatory.
+* Good, Industry Usage/Standard: React has a large and active community of developers, with many resources available, including libraries, tutorials, and support.
```
Comment (Vinnl) [inline docs/adr/0004-framework.md:43]: Related advantage I just realised: this makes it easier for drive-by contributors to contribute, since they can use existing familiarity with React that they wouldn't have with our homegrown snowflake framework. (We've had a number of contributions to Relay's UI since it's been in React, which is awesome.)

File: docs/adr/0004-framework.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+  * Accessibility: Access to [React Aria](https://react-spectrum.adobe.com/react-aria/) (Relay also uses this library).
+* Good, [fluent-react](https://github.com/projectfluent/fluent.js/tree/main/fluent-react) enables localised messages that can be updated on the client-side
+* Good, Avoid snowflake configs: we won't have to manually set up tooling and make sure they work well together, like [#3003](https://github.com/mozilla/blurts-server/pull/2987) and [#2987](https://github.com/mozilla/blurts-server/pull/2987)
+* Good, Automatic code splitting, avoiding e.g. cumulative layout shift, and decreating bundle sizes.
+* Good, Poor front-end developer experience: VSCode code plugins for code completeion, error linting, etc.
```
Comment (Vinnl) [inline docs/adr/0004-framework.md:48]:
```suggestion
* Good, Great front-end developer experience: VSCode code plugins for code completeion, error linting, hot reloading for rapid iteration, etc. 
```

Comment (maxxcrawford) [reply docs/adr/0004-framework.md:48]: @Vinnl 🙈 What a typo! Good suggestion.

File: docs/adr/0004-framework.md [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+* Good, Great front-end developer experience: VSCode code plugins for code completeion, error linting, hot reloading for rapid iteration, etc. 
+* Bad, Tooling Overhead: React's tooling and ecosystem can be complex, which can be a disadvantage for developers who prefer a more minimalistic approach.
+* Bad, Performance cost when rendering a large number of components, may require additional optimization work
+* Bad, Learning curve: Challenging for new developers jumping into the project to understand how to debug issues or how components are interacting with each other
+* Bad, Time Cost: Time spent porting existing site over to React
+* Bad, We give away some autonomy w.r.t. what tooling we can use and when we can use the latest versions.
+* Bad, The client-side execution time of React is additional overhead.
+* Bad, No portability or interoperability to non-React projects.
+* Bad, React will require maintenance and periodic upgrades
+* Bad, Rendering performance cost: Inital load of React
```
Comment (pdehaan) [inline docs/adr/0004-framework.md:57]:
```suggestion
* Bad, Rendering performance cost: Initial load of React 
```

File: docs/adr/0004-framework.md [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+* Good, Enables static analysis: programmatically-generated DOM and explicit module imports enables optimal use of TypeScript to catch errors at compile-time and make code more self-explanatory.
+* Good, Industry Usage/Standard: React has a large and active community of developers, with many resources available, including libraries, tutorials, and support.
+  * Accessibility: Access to [React Aria](https://react-spectrum.adobe.com/react-aria/) (Relay also uses this library).
+* Good, [fluent-react](https://github.com/projectfluent/fluent.js/tree/main/fluent-react) enables localised messages that can be updated on the client-side
+* Good, Avoid snowflake configs: we won't have to manually set up tooling and make sure they work well together, like [#3003](https://github.com/mozilla/blurts-server/pull/2987) and [#2987](https://github.com/mozilla/blurts-server/pull/2987)
+* Good, Automatic code splitting, avoiding e.g. cumulative layout shift, and decreating bundle sizes.
+* Good, Great front-end developer experience: VSCode code plugins for code completeion, error linting, hot reloading for rapid iteration, etc.
```
Comment (pdehaan) [inline docs/adr/0004-framework.md:48]:
```suggestion
* Good, Great front-end developer experience: VSCode code plugins for code completion, error linting, hot reloading for rapid iteration, etc. 
```

File: docs/adr/0004-framework.md [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+	* Note that FxA came to a similiar decision: [FxA ADR: Refactor Subscription Platform frontend - with Next.js](https://github.com/mozilla/fxa/blob/main/docs/adr/0035-refactor-payments-frontend-with-nextjs.md) 
+* Good, Enables static analysis: programmatically-generated DOM and explicit module imports enables optimal use of TypeScript to catch errors at compile-time and make code more self-explanatory.
+* Good, Industry Usage/Standard: React has a large and active community of developers, with many resources available, including libraries, tutorials, and support.
+  * Accessibility: Access to [React Aria](https://react-spectrum.adobe.com/react-aria/) (Relay also uses this library).
+* Good, [fluent-react](https://github.com/projectfluent/fluent.js/tree/main/fluent-react) enables localised messages that can be updated on the client-side
+* Good, Avoid snowflake configs: we won't have to manually set up tooling and make sure they work well together, like [#3003](https://github.com/mozilla/blurts-server/pull/2987) and [#2987](https://github.com/mozilla/blurts-server/pull/2987)
```
Comment (pdehaan) [inline docs/adr/0004-framework.md:46]:
```suggestion
* Good, Avoid snowflake configs: we won't have to manually set up tooling and make sure they work well together, like [#3003](https://github.com/mozilla/blurts-server/pull/3003) and [#2987](https://github.com/mozilla/blurts-server/pull/2987)
```

File: docs/adr/0004-framework.md [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+## Pros and Cons of the Options <!-- optional -->
+
+### Migrate framework to React and Next.js
+
+* Good, Component-based Architecture: This simplifies code maintenance by breaking complex UIs into reusable components.
+* Good, Testing: the virtual DOM makes it feasible to write non-brittle (i.e. without needing to fire up a full browser) unit tests of the front-end
+* Good, Fewer bugs: having React take care of removing, adding or changing DOM nodes removes a large surface area for bugs.
+* Good, Shared knowledge: This approach shares the same structure as Firefox Relay. This would immediately allow the team to reuse and reference.
+	* Note that FxA came to a similiar decision: [FxA ADR: Refactor Subscription Platform frontend - with Next.js](https://github.com/mozilla/fxa/blob/main/docs/adr/0035-refactor-payments-frontend-with-nextjs.md)
```
Comment (pdehaan) [inline docs/adr/0004-framework.md:41]:
```suggestion
	* Note that FxA came to a similar decision: [FxA ADR: Refactor Subscription Platform frontend - with Next.js](https://github.com/mozilla/fxa/blob/main/docs/adr/0035-refactor-payments-frontend-with-nextjs.md) 
```

File: docs/adr/0004-framework.md [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+* Bad, The client-side execution time of React is additional overhead.
+* Bad, No portability or interoperability to non-React projects.
+* Bad, React will require maintenance and periodic upgrades
+* Bad, Rendering performance cost: Inital load of React 
+
+## (Secondary Decision) Migration Plan Options
+
+There are roughly three options to migrate to React
+
+#### 1 Static-site generation (SSG)
```
Comment (pdehaan) [inline docs/adr/0004-framework.md:63]: Adding period for consistency with items 2+3 below.
```suggestion
#### 1. Static-site generation (SSG)
```

File: docs/adr/0004-framework.md (hunk 2 above)
Comment (maxxcrawford) [reply docs/adr/0004-framework.md:51]: Added in 7b9456b5db28ad6cd00ee742d76295948c346429

**Comment by maxxcrawford** [comment]
I merged this decision as decided. Note that we have some follow-on decision to document further.

File: docs/adr/0004-framework.md (hunk 3 above)
Comment (pdehaan) [reply docs/adr/0004-framework.md:1]: @maxxcrawford Not sure if you still want to rename the ADR, post merge. 🤷

Comment (maxxcrawford) [reply docs/adr/0004-framework.md:1]: Renamed!
